socorro/external/boto/connection_context.py

```diff
@@ -0,0 +1,245 @@
+    )
+
+    conditional_exceptions = (
+        boto.exception.StorageResponseError
+    )
+
+    #--------------------------------------------------------------------------
+    def is_operational_exception(self, x):
+        if "not found, no value returned" in str(x):
+            # the not found error needs to be re-tryable to compensate for
+            # eventual consistency.  However, a method capable of raising this
+            # exception should never be used with a transaction executor that
+            # has infinite back off.
+            return True
+        #elif   # for further cases...
+        return False
+
+
+    #--------------------------------------------------------------------------
+    def __init__(self, config, quit_check_callback=None):
+        self.config =  config
+
+        self._connect_to_endpoint = boto.connect_s3
+        self._calling_format = config.calling_format
+        self._CreateError = boto.exception.S3CreateError
+        self._S3ResponseError = boto.exception.S3ResponseError
+        self._open = open
+
+
+    #--------------------------------------------------------------------------
+    @staticmethod
+    def build_s3_dirs(prefix, name_of_thing, id):
+        """
+        Use S3 pseudo-directories to make it easier to list/expire later
+        {{prefix}}/{{version}}/{{name_of_thing}}/{{id}}
+        """
+        version = 'v1'
+        return '%s/%s/%s/%s' % (prefix, version, name_of_thing, id)
+
+    #--------------------------------------------------------------------------
+    def _get_bucket(self, conn, bucket_name):
+        try:
+            return self._bucket_cache
+        except AttributeError:
+            self._bucket_cache = conn.get_bucket(bucket_name)
+            return self._bucket_cache
+
+    #--------------------------------------------------------------------------
+    def _get_or_create_bucket(self, conn, bucket_name):
+        try:
+            return self._bucket_cache
+        except AttributeError:
+            try:
+                self._bucket_cache = conn.get_bucket(bucket_name)
+            except self._S3ResponseError:
+                self._bucket_cache = conn.create_bucket(bucket_name)
+            return self._bucket_cache
+
+    #--------------------------------------------------------------------------
+    def submit_to_boto_s3(self, id, name_of_thing, thing):
+        """submit something to boto.
+        """
+        if not isinstance(thing, basestring):
+            raise Exception('can only submit strings to boto')
+
+        conn = self._connect()
+        bucket = self._get_or_create_bucket(conn, self.config.bucket_name)
+
+        key = self.build_s3_dirs(self.config.prefix, name_of_thing, id)
+
+        storage_key = bucket.new_key(key)
+        storage_key.set_contents_from_string(thing)
+
+    #--------------------------------------------------------------------------
+    def fetch_from_boto_s3(self, id, name_of_thing):
```

Comment:
Or `get_from_boto` considering that it's always supposed to return the content or raise an exception. To be "fetch" is an activity you can do in the background or something where you don't need the reading.